lib/charms/sdcore_upf/v0/fiveg_n4.py

```diff
@@ -0,0 +1,290 @@
+
+    ```python
+    ...
+    from charms.sdcore_upf.v0.fiveg_n4 import N4Provides
+    ...
+
+    class SomeProviderCharm(CharmBase):
+
+        def __init__(self, *args):
+            ...
+            self.fiveg_n4 = N4Provides(charm=self, relation_name="fiveg_n4")
+            ...
+            self.framework.observe(self.fiveg_n4.on.fiveg_n4_request, self._on_fiveg_n4_request)
+
+        def _on_fiveg_n4_request(self, event):
+            ...
+            self.fiveg_n4.publish_upf_n4_information(
+                relation_id=event.relation_id,
+                upf_hostname=hostname,
+                upf_port=n4_port,
+            )
+    ```
+
+    And a corresponding section in charm's `metadata.yaml`:
+    ```
+    provides:
+        fiveg_n4:  # Relation name
+            interface: fiveg_n4  # Relation interface
+    ```
+
+Charms that require the `fiveg_n4` relation should use `N4Requires`.
+Typical usage of this class would look something like:
+
+    ```python
+    ...
+    from charms.sdcore_upf.v0.fiveg_n4 import N4Requires
+    ...
+
+    class SomeRequirerCharm(CharmBase):
+
+        def __init__(self, *args):
+            ...
+            self.fiveg_n4 = N4Requires(charm=self, relation_name="fiveg_n4")
+            ...
+            self.framework.observe(self.upf.on.fiveg_n4_available, self._on_fiveg_n4_available)
+
+        def _on_fiveg_n4_available(self, event):
+            upf_hostname = event.upf_hostname,
+            upf_port = event.upf_port,
+            # Do something with the UPF's hostname and port
+    ```
+
+    And a corresponding section in charm's `metadata.yaml`:
+    ```
+    requires:
+        fiveg_n4:  # Relation name
+            interface: fiveg_n4  # Relation interface
+    ```
+"""
+
+import logging
+
+from interface_tester.schema_base import DataBagSchema  # type: ignore[import]
+from ops.charm import CharmBase, CharmEvents, RelationChangedEvent, RelationJoinedEvent
+from ops.framework import EventBase, EventSource, Object
+from pydantic import BaseModel, Field, ValidationError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "bc6261cf77104d4fb3edfd6d4ea63149"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = ["pydantic", "pytest-interface-tester"]
+
+
+logger = logging.getLogger(__name__)
+
+"""Schemas definition for the provider and requirer sides of the `fiveg_n3` interface.
+It exposes two interfaces.schema_base.DataBagSchema subclasses called:
+- ProviderSchema
+- RequirerSchema
+Examples:
+    ProviderSchema:
+        unit: <empty>
+        app: {
+            "upf_hostname": "upf.uplane-cloud.canonical.com",
+            "upf_port": 8805
+        }
+    RequirerSchema:
+        unit: <empty>
+        app:  <empty>
+"""
+
+
+class ProviderAppData(BaseModel):
```

Comment:
I guess it doesn't really matter, because one has no connection to the other, but it sounds like a good idea to have the same name in both places, so I changed it.
Thanks!